clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -290,18 +317,47 @@ public void poll(long timeout) throws IOException {
 
         clear();
 
-        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty())
+        boolean dataInBuffers = !keysWithBytesBuffered.isEmpty();
+        
+        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty() || (madeProgressLastPoll && dataInBuffers))
             timeout = 0;
 
+        if (!memoryPool.isOutOfMemory() && (outOfMemory || dataInBuffers)) {
```

Comment:
youre probably right. if dataInBuffers = true it means either:

1. there is data in app buffer. only way (i think?) to get to this situation is that it could not be read out of app buffer because no memory, hence outOfMemory will be true, which will be enough to trigger an unmute when memory becomes available

2. there's data only in net buffer. this means must data must come from socket and we have successfully read out everything that may have been in app buffer, so we didnt run OOM, so channel is not muted and will show up in a future poll as a read key